Build pipeline html parsing cv

cvalarida opened the pull request:
## Description
This PR speeds up the Metalsmith build by parsing all HTML files only once.
- Adds a plugin to the Metalsmith pipeline that parses out all the html files
- Passes that parsed object along with each Metalsmith file
- Modifies the four plugins that were each individually parsing the files to use the new `parsedContent`
- Modifies those plugins to also _not_ write to `file.contents`, but instead add a `modified` flag when they would normally have written
- Adds a plugin that rewrites `file.contents` one time for each file that has the `modified` flag

## Testing done
Our Metalsmith build adds random IDs to elements with an `[onclick]` handler. This means we can't just checksum the build output to ensure that nothing has changed.

My testing process was like this:
**Test the nonce script plugin**
1) Build
2) Output a list of MD5 checksums for every file in the build
    - `find build/localhost/ -type f -exec md5sum {} + | awk '{print $2 "  " $1}' | sort > first-build`
3) Build again
4) Output another list of MD5 checksums into a new file
    - `find build/localhost/ -type f -exec md5sum {} + | awk '{print $2 "  " $1}' | sort > second-build`
5) `diff`ed the two to see which files changed
6) Check the differences; this shows the files that have changed by the nonce script plugin
7) Modify the build pipeline
8) Output another list of MD5 checksums into another new file
    - `find build/localhost/ -type f -exec md5sum {} + | awk '{print $2 "  " $1}' | sort > after-modification`
9) Diff the output
    - Unfortunately, this shows a _lot_ more modifications because the nonce script plugin now uses Cheerio instead of JSDOM, so the formatting is somewhat different
    ![image](https://user-images.githubusercontent.com/12970166/65629421-2755ee00-df88-11e9-989f-2f82e67293f6.png)
10) Check files that are expected to be different (from step 6)
    - The expected differences show up
    - `nonce` is added to inline scripts and not to scripts with a `src` attribute

**Check the update links plugin**
This one adds `rel="noopener"`.
1) Build the baseline
2) `mv build/localhost build/localhost-before`
3) Make the modifications
4) Check the count on the build before modification
    - `find build/localhost-before/ -name '*.html' -exec grep 'noopener' {} + | wc -l`
    - 1504
5) Check the count on the build after modification
    - `find build/localhost/ -name '*.html' -exec grep 'noopener' {} + | wc -l`
    - 1504

**Check the IDs**
1) Modify the plugin to output the files which were modified
2) Build
3) Open a file from the list of modified files
4) Find an `h2` or `h3` and ensure that the IDs are there
    - From what I could tell, they are

**Check the broken links plugin**
I had broken links locally for missing PDFs. After modifying the build, those same broken links were still reported.

**Reordering the plugins**
While testing, I found that a couple problems with keeping the previous plugin order:
- Parsing before the nonce plugin resulted in an error where the Drupal debug page created a new HTML file that wasn't parsed
- Putting the "redirect va.gov links" plugin before one that used `parsedContent` overrode the results of those redirects because the changes weren't made in the parsed content, just `file.contents`
With that in mind, I grouped all the plugins that require the parsed content together.

## Screenshots
![image](https://user-images.githubusercontent.com/12970166/65632946-c7634580-df8f-11e9-8ec8-003f34863974.png)

## Acceptance criteria
- [x] The HTML files are parsed only once
- [x] All the plugins still function as expected
- [x] The result of the other plugins is not affected

Commit:
Add parseHtml plugin and use it; reorder functions that parse the HTML

Commit:
Use parsedContent in three plugins

Commit:
Rewrite nonce script plugin to use cheerio (parsedContent)

Commit:
Fix adding the ID in the nonce script

Commit:
Fix adding nonce attribute

Commit:
Clean up comment

Commit:
Use outputHtml plugin so we write the contents only once

cvalarida commented on src/site/stages/build/plugins/add-nonce-to-scripts.js:
```diff
@@ -1,4 +1,4 @@
-const jsdom = require('jsdom');
```
This file had to change quite a bit because jsdom does things a little differently from cheerio.

ncksllvn commented on src/site/stages/build/plugins/parse-html.js:
```diff
@@ -0,0 +1,13 @@
+const parseHtml = files => {
+  for (const fileName of Object.keys(files)) {
+    if (path.extname(fileName) === '.html') {
+      files[fileName].parsedContent = cheerio.load(files[fileName].contents);
```
I like `parsedContent` but `file.$` or `file.dom` could be kinda snazzy

rianfowler replied:
+1 for file.dom

cvalarida replied:
`file.dom` it is! I want it to be obvious by just reading it, and `file.$` doesn't give much away. I like `file.dom`. :+1:

ncksllvn commented on src/site/stages/build/plugins/output-html.js:
```diff
@@ -0,0 +1,12 @@
+  }
+};
+
+module.exports = outputHtml;
```
Instead of `output-html`, maybe something like `finalize-html`? Or something verbose like `replace-file-contents-with-dom` 🤷‍♂

ncksllvn reviewed (approved):
Total nitpicks/naming suggestions - looks great!

Commit:
Use parsedContent in broken link unit test

Commit:
Expect one broken link

Previously, the HTML it was validating _had_ only one link, but with
the restructure, it's rendering the ${anchor} as well.

Commit:
Use file.dom

Commit:
Use file.dom in the broken link checker unit test too...

Commit:
Rename ouputHtml -> replaceContentsWithDom
